R/read-write.R

```diff
@@ -0,0 +1,465 @@
+  )
+  if (!all(special_files %in% file_list)) {
+    abort_files_or_sheets_missing(setdiff(special_files, intersect(file_list, special_files)), csv_directory)
+  }
+
+  def_base <- readr::read_csv(
+    file.path(csv_directory, "___info_file_dm.csv"),
+    col_types = "ccc"
+  )
+  table_names <- def_base$table
+  col_class_info_raw <- readr::read_csv(
+    file.path(csv_directory, "___coltypes_file_dm.csv"),
+    col_types = "ccc"
+  )
+  if (!all(col_class_info_raw$class %in% class_translation_r_readr$r_class)) {
+    abort_class_not_supported(setdiff(col_class_info_raw$class, class_translation_r_readr$r_class))
+  }
+  col_class_info <- col_class_info_raw %>%
+    mutate(readr_class = readr_translate(class)) %>%
+    group_by(table) %>%
+    summarize(readr_class = paste0(readr_class, collapse = ""))
+  # sort column classes according to column def_base$table
+  col_class_sorted <- col_class_info$readr_class[order(match(col_class_info$table, table_names))]
+
+  pk_info <- readr::read_csv(
+    file.path(csv_directory, "___pk_file_dm.csv"),
+    col_types = "cc"
+  )
+
+  fk_info <- readr::read_csv(
+    file.path(csv_directory, "___fk_file_dm.csv"),
+    col_types = "ccci"
+  )
+
+  table_files <- setdiff(
+    list.files(csv_directory),
+    c("___info_file_dm.csv", "___pk_file_dm.csv", "___fk_file_dm.csv", "___coltypes_file_dm.csv")
+  )
+  # sort table_files according to column def_base$table
+  table_files_sorted <- table_files[order(match(table_files, paste0(table_names, ".csv")))]
+
+  table_tibble <- map2(
+    file.path(csv_directory, table_files_sorted),
+    col_class_sorted,
+    ~ readr::read_csv(file = .x, col_types = .y)
+  ) %>%
+    set_names(table_names) %>%
+    enframe(name = "table", value = "data") %>%
+    # empty subset is necessary because
+    # https://www.tidyverse.org/blog/2018/12/readr-1-3-1/#tibble-subclass
+    mutate(data = map(data, `[`))
+
+  make_dm(def_base, table_tibble, pk_info, fk_info)
+}
+
+#' @inheritParams dm_write_csv
+#' @param zip_file_path
+#' For `dm_write_zip`: The file path to the `zip`-file to write.
+#'
+#' For `dm_read_zip`: The file path to the `zip`-file to read the `dm` from.
+#' @param overwrite Logical, default: `FALSE`. In case the file already exists, should it be overwritten?
+#' @details `dm_write_zip()`: write a `dm` to a `zip`-file containing a collection of `csv`-files.
+#' @rdname dm-read-write
+#' @export
+dm_write_zip <- function(dm, zip_file_path, overwrite = FALSE) {
+  if (file.exists(zip_file_path)) {
+    if (overwrite) {
+      message(glue::glue("Overwriting file {tick(zip_file_path)}."))
+    } else{
+      abort_file_exists(zip_file_path)
+    }
+  }
+
+  csv_directory <- file.path(tempdir(), basename(tempfile(pattern = "dm_zip_")))
+  dm_write_csv_impl(dm, csv_directory, zip = TRUE)
+
+  withr::defer(try(unlink(csv_directory, recursive = TRUE)))
+
+  csv_files <- list.files(csv_directory)
+  # compress the file ("-j" junks the path to the file)
+
+  zip(
```

Comment:
I suspect {zip} might work better here: https://cran.r-project.org/web/packages/zip/index.html.

Reply:
maybe if it works it's more platform agnostic, but in the first test (which works with `utils::zip()`), I get the error:
```r
Error in zip_internal(zipfile, files, recurse, compression_level, append = FALSE,  : 
  zip error: `Cannot add file `/var/folders/x3/ndmkxk1j2wn0mx2pw9v9httm0000gn/T//RtmpmKOSNA/dm_zip_1149248ed84b/___coltypes_file_dm.csv` to archive `___test_path/dm.zip`` in file `zip.c:348`
```
Not sure if it's worth the effort to try finding the source of this problem.